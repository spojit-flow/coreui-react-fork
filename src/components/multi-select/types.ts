export type Option = {
  custom?: boolean
  disabled?: boolean
  selected?: boolean
  text: string // TODO: change to label in v5
  value: number | string
  [key: string]: number | string | any
}

export type OptionsGroup = {
  label: string
  options?: Option[]
  [key: string]: number | string | any
}

export type SelectedOption = {
  disabled?: boolean
  text: string
  value: number | string
  [key: string]: number | string | any
}
