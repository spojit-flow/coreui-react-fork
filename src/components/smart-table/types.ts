import { ReactNode } from 'react'

import { CTableDataCellProps } from '../table/CTableDataCell'
import { CTableHeaderCellProps } from '../table/CTableHeaderCell'
import { CTableRowProps } from '../table/CTableRow'

export type Column = {
  children?: Column[]
  filter?: boolean | ((values: any[], onChange: (value: any) => void) => ReactNode)
  group?: string
  key: string
  label?: string
  sorter?: boolean
  _style?: any
  _props?: CTableHeaderCellProps
}

export type ColumnFilter = {
  lazy?: boolean
  external?: boolean
}

export type ColumnFilterValue = {
  [key: string]: any
}

export type FooterItem = {
  label?: string
  _props?: CTableDataCellProps
}

export type Group = {
  children?: Group[] | Column[]
  colspan?: number
  deep?: number
  group?: string
  key: string
  label?: string
  _style?: any
  _props?: CTableHeaderCellProps
}

export type Item = {
  [key: string]: number | string | any
  _cellProps?: any
  _props?: CTableRowProps
  _selected?: boolean
}

export type ItemsPerPageSelect = {
  external?: boolean
  label?: string
  values?: Array<number>
}

export type Pagination = {
  external?: boolean
}

export type Sorter = {
  resetable?: boolean
  external?: boolean
}

export type ScopedColumns = {
  [key: string]: any
  details?: (a: Item, b: number) => ReactNode
}

export type SorterValue = {
  column?: string
  state?: number | string
}

export type TableFilter = {
  lazy?: boolean
  external?: boolean
}
