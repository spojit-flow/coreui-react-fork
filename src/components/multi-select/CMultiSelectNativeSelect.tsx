import React, { forwardRef, InputHTMLAttributes } from 'react'
import PropTypes from 'prop-types'

import type { Option } from './types'

export interface CMultiSelectNativeSelectProps
  extends Omit<InputHTMLAttributes<HTMLSelectElement>, 'options'> {
  options?: Option[]
  value?: string | number | string[]
}

const createNativeOptions = (options: Option[]) =>
  options &&
  options.map((option: Option, index: number) =>
    option.options ? (
      <optgroup label={option.label} key={index}>
        {createNativeOptions(option.options)}
      </optgroup>
    ) : (
      <option value={option.value} key={index} disabled={option.disabled}>
        {option.text}
      </option>
    ),
  )

export const CMultiSelectNativeSelect = forwardRef<
  HTMLSelectElement,
  CMultiSelectNativeSelectProps
>(({ id, options, ...rest }, ref) => {
  return (
    <select
      className="multi-select-new"
      {...(id && { id: `${id}-multi-select` })}
      {...(id && { name: `${id}-multi-select` })}
      tabIndex={-1}
      style={{ display: 'none' }}
      {...rest}
      ref={ref}
    >
      {options && createNativeOptions(options)}
    </select>
  )
})

CMultiSelectNativeSelect.propTypes = {
  options: PropTypes.array,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string.isRequired),
  ]),
}

CMultiSelectNativeSelect.displayName = 'CMultiSelectNativeSelect'
