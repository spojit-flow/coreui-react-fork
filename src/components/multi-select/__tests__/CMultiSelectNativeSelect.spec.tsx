import * as React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { CMultiSelectNativeSelect } from '../CMultiSelectNativeSelect'
import { Option } from '../types'

test('loads and displays CMultiSelectNativeSelect component', async () => {
  const { container } = render(<CMultiSelectNativeSelect options={[]} />)
  expect(container).toMatchSnapshot()
})

test('CMultiSelectNativeSelect customize', async () => {
  const options: Option[] = [
    {
      text: 'text1',
      value: 'value1',
    },
    {
      text: 'text2',
      value: 'value2',
      disabled: true,
      label: 'label2',
      order: 4,
      selected: true,
    },
    {
      text: 'text3',
      value: 'value3',
      order: 3,
    },
  ]
  const { container } = render(
    <CMultiSelectNativeSelect multiple={true} options={options} value="3" />,
  )
  expect(container).toMatchSnapshot()
})
