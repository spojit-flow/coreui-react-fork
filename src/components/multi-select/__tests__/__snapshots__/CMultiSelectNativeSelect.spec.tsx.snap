// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMultiSelectNativeSelect customize 1`] = `
<div>
  <select
    class="multi-select-new"
    multiple=""
    style="display: none;"
    tabindex="-1"
  >
    <option
      value="value1"
    >
      text1
    </option>
    <option
      disabled=""
      value="value2"
    >
      text2
    </option>
    <option
      value="value3"
    >
      text3
    </option>
  </select>
</div>
`;

exports[`loads and displays CMultiSelectNativeSelect component 1`] = `
<div>
  <select
    class="multi-select-new"
    style="display: none;"
    tabindex="-1"
  />
</div>
`;
